x/feepay/ante/deduct_fee.go

```diff
@@ -0,0 +1,269 @@
+package ante
+
+import (
+	"fmt"
+	"math"
+
+	wasmtypes "github.com/CosmWasm/wasmd/x/wasm/types"
+
+	errorsmod "cosmossdk.io/errors"
+	sdkmath "cosmossdk.io/math"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/cosmos-sdk/x/auth/ante"
+	"github.com/cosmos/cosmos-sdk/x/auth/types"
+	bankkeeper "github.com/cosmos/cosmos-sdk/x/bank/keeper"
+
+	feepayhelpers "github.com/CosmosContracts/juno/v18/x/feepay/helpers"
+	feepaykeeper "github.com/CosmosContracts/juno/v18/x/feepay/keeper"
+	feepaytypes "github.com/CosmosContracts/juno/v18/x/feepay/types"
+	globalfeekeeper "github.com/CosmosContracts/juno/v18/x/globalfee/keeper"
+)
+
+// DeductFeeDecorator deducts fees from the first signer of the tx
+// If the first signer does not have the funds to pay for the fees, return with InsufficientFunds error
+// Call next AnteHandler if fees successfully deducted
+// CONTRACT: Tx must implement FeeTx interface to use DeductFeeDecorator
+//
+// Additionally, the Deduct Fee ante is a fork of the SDK's DeductFeeDecorator. This decorator looks for single
+// message transactions with no provided fee. If they correspond to a registered FeePay Contract, the FeePay
+// module will cover the cost of the fee (if the balance permits).
+type DeductFeeDecorator struct {
+	feepayKeeper    feepaykeeper.Keeper
+	globalfeeKeeper globalfeekeeper.Keeper
+	accountKeeper   ante.AccountKeeper
+	bankKeeper      bankkeeper.Keeper
+	feegrantKeeper  ante.FeegrantKeeper
+	bondDenom       string
+}
+
+func NewDeductFeeDecorator(fpk feepaykeeper.Keeper, gfk globalfeekeeper.Keeper, ak ante.AccountKeeper, bk bankkeeper.Keeper, fgk ante.FeegrantKeeper, tfc ante.TxFeeChecker, bondDenom string) DeductFeeDecorator {
```

Comment:
Does the ante.TxFeeChecker still need to be set? since this was forked from the SDK I assume so? does not seem to be used in GlobalFee afaik

Reply:
No, we do not want the logic to be overridden by the TxFeeChecker so we define it in the ante itself.

Reply:
I removed the unnecessary parameter.